src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,168 @@
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        return node(index).item;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> node = node(index);
+        node.item = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        return unlink(node(index));
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        Node<T> node;
+        T value = null;
+        if (t == null) {
+            for (node = head; node != null; node = node.next) {
+                if (node.item == null) {
+                    value = unlink(node);
+                }
+            }
+        } else {
+            for (node = head; node != null; node = node.next) {
+                if (t.equals(node.item)) {
+                    value = unlink(node);
+                }
+            }
+        }
+        return value;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void linkToEnd(T t) {
+        Node<T> l = tail;
+        Node<T> node = new Node<T>(t, null, l);
+        tail = node;
+        if (l == null) {
+            head = node;
+        } else {
+            l.next = node;
+        }
+        ++size;
+    }
+
+    private Node<T> node(int index) {
+        Node<T> node = head;
+        for (int i = 0; i < index; ++i) {
+            node = node.next;
+        }
+        return node;
+    }
+
+    private void linkBefore(T t, Node<T> src) {
+        Node<T> pred = src.prev;
+        Node<T> node = new Node<T>(t, src, pred);
+        src.prev = node;
+        if (pred == null) {
+            head = node;
+        } else {
+            pred.next = node;
+        }
+        ++size;
+    }
+
+    private T unlink(Node<T> node) {
+        Node<T> next = node.next;
+        Node<T> prev = node.prev;
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        T element = node.item;
```

Comment:
we can write:
node = null
and then remove `node.prev = null;` and `node.next = null;`